### PR TITLE
pbss: support to load legacy async buffer format

### PR DESCRIPTION
**Pull request by galaio**
### Description

This PR fixes a legacy journal format issue, it tries to load an async buffer in a legacy journal file.

When the PR is enabled, the node can read the async buffer correctly.

```bash
logs/bsc.log.2024-12-20_11:t=2024-12-20T11:15:10+0000 lvl=info msg="New async node buffer" limit="256.00 MiB" layers=44221
logs/bsc.log.2024-12-20_12:t=2024-12-20T12:15:37+0000 lvl=info msg="Failed to load journal, discard it" err="unexpected journal version want 2 got 0"
```

It will fix the issue https://github.com/bnb-chain/bsc/issues/2796.

### Changes

Notable changes: 
* pbss: support to load legacy async buffer format;
* ...

### CODE REVIEW

**Review by Fart10 — APPROVED** [review]
..

File: triedb/pathdb/journal.go [hunk 1]
```diff
@@ -207,7 +207,7 @@ func (db *Database) loadJournal(diskRoot common.Hash) (layer, error) {
 		return nil, fmt.Errorf("%w want %x got %x", errUnmatchedJournal, root, diskRoot)
 	}
 	// Load the disk layer from the journal
-	base, err := db.loadDiskLayer(r, journalTypeForReader)
+	base, err := db.loadDiskLayer(r, journalTypeForReader, version)
```
Comment (zzzckck) [inline triedb/pathdb/journal.go:210]: but previous Line#195 always returned due to version mismatch

Comment (zzzckck) [reply triedb/pathdb/journal.go:210]: > but previous Line#195 always returned due to version mismatch

will try to load asyncbuff after failed to load journal

**Comment by galaio** [comment]
PBSS journal history change:
1. journal ver1, https://github.com/ethereum/go-ethereum/pull/28940
2. journal ver2, https://github.com/ethereum/go-ethereum/pull/30643

The async buffer saved in the journal only changes in ver2, so we need to handle the legacy async buffer ver0/ver1 in the same way.

File: triedb/pathdb/journal.go [hunk 2]
```diff
@@ -238,13 +238,52 @@ func (db *Database) loadLayers() layer {
 	if !(root == types.EmptyRootHash && errors.Is(err, errMissJournal)) {
 		log.Info("Failed to load journal, discard it", "err", err)
 	}
+	// try to load async buffer only, it can be compatible with old journal ver
+	base, err := db.loadAsyncBufferAsJournalV0V1(root)
```
Comment (buddh0) [inline triedb/pathdb/journal.go:242]: The method of journaling is independent of the buffer type.

**Review by zzzckck — APPROVED** [review]

File: triedb/pathdb/journal.go [hunk 3]
```diff
@@ -238,13 +238,53 @@ func (db *Database) loadLayers() layer {
...
+	if err == nil {
+		log.Info("load legacy node buffer successful", "new root", base.rootHash(), "disk root", root)
+		return base
+	}
+	log.Warn("Failed to load disk journal with node buffer, discard it", "err", err)
 	// Return single layer with persistent state.
 	return newDiskLayer(root, rawdb.ReadPersistentStateID(db.diskdb), db, nil, NewTrieNodeBuffer(db.config.SyncFlush, db.config.WriteBufferSize, nil, nil, 0))
 }
 
+// loadNodeBufferAsJournalV0V1 try to load legacy node buffer data from journal
```
Comment (NathanBSC) [inline triedb/pathdb/journal.go:252]: leave a todo here: after prague or one more hardfork, this PR can be reverted

File: triedb/pathdb/buffer.go [hunk 4]
```diff
@@ -95,7 +95,12 @@ func (b *buffer) revertTo(db ethdb.KeyValueReader, nodes map[common.Hash]map[str
 		return nil
 	}
 	b.nodes.revertTo(db, nodes)
-	b.states.revertTo(accounts, storages)
+	// TODO(galaio): In order to be compatible with the legacy version, a temporary empty check is added,
+	// which may affect the reading result of pbss as flatReader, see: flatReader.Account()
+	// it could be removed in the future
+	if len(b.states.accountData) != 0 || len(b.states.storageData) != 0 {
```
Comment (NathanBSC) [inline triedb/pathdb/buffer.go:101]: need add one more scenario which will lead to panic

**Review by buddh0 — APPROVED** [review]

**Review by zzzckck — APPROVED** [review]

**Review by zzzckck — APPROVED** [review]

File: triedb/pathdb/journal.go (hunk 3 above)
Comment (Fart10) [reply triedb/pathdb/journal.go:252]: yes

File: triedb/pathdb/buffer.go (hunk 4 above)
Comment (Fart10) [reply triedb/pathdb/buffer.go:102]: 👍

**Review by Fart10 — COMMENTED** [review]
..

**Pull request merged**